tests/Unit/Evolution/Systems/Ccz4/Test_DerivZ4Constraint.cpp

```diff
@@ -0,0 +1,40 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Framework/TestingFramework.hpp"
+
+#include <cstddef>
+#include <string>
+
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Evolution/Systems/Ccz4/DerivZ4Constraint.hpp"
+#include "Framework/CheckWithRandomValues.hpp"
+#include "Framework/SetupLocalPythonEnvironment.hpp"
+#include "Framework/TestHelpers.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace {
+template <size_t Dim, typename DataType>
+void test_compute_grad_spatial_z4_constraint(const DataType& used_for_size) {
+  pypp::check_with_random_values<1>(
+      static_cast<tnsr::ij<DataType, Dim, Frame::Inertial> (*)(
+          const tnsr::i<DataType, Dim, Frame::Inertial>&,
+          const tnsr::ii<DataType, Dim, Frame::Inertial>&,
+          const tnsr::Ijj<DataType, Dim, Frame::Inertial>&,
+          const tnsr::ijj<DataType, Dim, Frame::Inertial>&,
+          const tnsr::I<DataType, Dim, Frame::Inertial>&,
+          const tnsr::iJ<DataType, Dim, Frame::Inertial>&)>(
+          &Ccz4::grad_spatial_z4_constraint<Dim, Frame::Inertial, DataType>),
+      "DerivZ4Constraint", "grad_spatial_z4_constraint", {{{-1., 1.}}},
+      used_for_size);
+}
+}  // namespace
+
+SPECTRE_TEST_CASE("Unit.Evolution.Systems.Ccz4.DerivZ4Constraint",
+                  "[Unit][Evolution]") {
+  pypp::SetupLocalPythonEnvironment local_python_env("Evolution/Systems/Ccz4/");
+
+  GENERATE_UNINITIALIZED_DOUBLE_AND_DATAVECTOR;
+  CHECK_FOR_DOUBLES_AND_DATAVECTORS(test_compute_grad_spatial_z4_constraint,
```

Comment:
Is there some reasonably easy way to verify that for KS with hat{Gamma}==tilde{Gamma} the result is zero? If not, can you at least add a test that verifies that if hat{Gamma}==tilde{Gamma} and Z_l=0. that the result is zero? I realize it's unlikely to have this bug, but it's a consistency check that for any D, tilde{gamma} and Gamma the constraint can vanish

Reply:
I think to be able to do the former with KS, it would first require being able to compute `\hat{\Gamma}` with KS but without needing `Z_i` as input, or to compute `Z_i` without `\hat{\Gamma}` as input. In the paper, the only definition I see depends on `Z_i`. Is there another way to compute either `Z_i` or `\hat{\Gamma}`? This may be a dumb question, since it seems the point here is that *because* `Z_i` is a constraint, these two tensors are defined in terms of each other. Just wanted to ask to be sure.

If not, I think that for the latter, two possible forms the test could take:

(1) pick a random `\tilde{\Gamma}^i`, initialize `\hat{\Gamma}^i` to be the same, compute their difference with for loops (should be 0), compute the numerical derivative (same for both), compute their difference with for loops (should be 0), compute `Z_i` from `\tilde{\Gamma}^i` and `\hat{\Gamma}^i` (should be 0), feed these inputs into `grad_spatial_z4_constraint` and make sure the result is 0

My issue with this the computation of the `tilde - hat` difference, its numerical derivative, and `Z_i` kind of feels pointless. If so, another option:

(2) Similar to (1), but just set the difference and the numerical derivative of the difference to 0

Let me know what you think. I don't feel great about either option.

Reply:
I think what makes this consistency check currently hard to test is that the function takes the `tilde -hat` difference and its derivative as arguments